### PR TITLE
[libc++] add floating point type check for uniform real distribution

### PR DESCRIPTION
**Pull request by changkhothuychung**
Created a PR to fix this issue - https://github.com/llvm/llvm-project/issues/62433

### CODE REVIEW

**Comment by NhatNguyen1810** [comment]
Can anyone tell me what is the issue with the Modular Build? I dont quite understand. How should I fix this?

**Comment by tschuett** [comment]
The error message says that there is a missing include for `std::is_floating_point`.

**Comment by NhatNguyen1810** [comment]
@philnik777 phik
can you let me know how I can fix the test file? I dont understand the error log and have no idea how to fix.

**Comment by AdvenamTacet** [comment]
@NhatNguyen1810 could you add a better description to the test? I looked at that file first and wasn't sure what's the purpose without reading the whole test carefully.

**Comment by NhatNguyen1810** [comment]
> @NhatNguyen1810 could you add a better description to the test? I looked at that file first and wasn't sure what's the purpose without reading the whole test carefully.

I added more descriptions to the test file.

**Comment by NhatNguyen1810** [comment]
@philnik777 can you review when you have time and let me know if it looks good to be approved? There is a workflow awaiting approval, can you run it for me? 

Thanks!

**Review by philnik777 — CHANGES_REQUESTED** [review]
Thanks for working on this! Some minor comments.

**Comment by NhatNguyen1810** [comment]
@philnik777 phik
can you take a quick look with my change? Basically I follow this file to do the regex matcher. I can't find a similar case where it does not use the regex. Let me know what I should put without the regex matcher. Thanks. 

https://github.com/llvm/llvm-project/blob/main/libcxx/test/libcxx/atomics/atomics.types.operations/atomics.types.operations.req/atomic_fetch_add.verify.cpp#L49

File: libcxx/include/__random/uniform_real_distribution.h
Comment (NhatNguyen1810) [reply libcxx/include/__random/uniform_real_distribution.h]: fixed

File: libcxx/include/__random/uniform_real_distribution.h
Comment (NhatNguyen1810) [reply libcxx/include/__random/uniform_real_distribution.h]: fixed

File: libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp
Comment (NhatNguyen1810) [reply libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp]: can you take a quick look with my change? Basically I follow this file to do the regex matcher. I can't find a similar case where it does not use the regex. Let me know what I should put without the regex matcher. Thanks.

https://github.com/llvm/llvm-project/blob/main/libcxx/test/libcxx/atomics/atomics.types.operations/atomics.types.operations.req/atomic_fetch_add.verify.cpp#L49

Or you can just make a change suggestion without regex which works, I will apply it.

Comment (NhatNguyen1810) [reply libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp]: @philnik777 it passed now. let me know if it looks good

**Comment by ldionne** [comment]
We have a utility called `__libcpp_random_is_valid_urng` in [`libcxx/include/__random/is_valid.h`](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__random/is_valid.h#L57-L61) -- should we instead create something similar to it and handle http://eel.is/c++draft/rand.req.genl more generally?

**Comment by ldionne** [comment]
> @ldionne Idio can you let me know which error the failed CI is about? And how should I fix it?

The CI looks green: https://buildkite.com/llvm-project/github-pull-requests/builds/11803

File: libcxx/utils/data/ignore_format.txt
Comment (ldionne) [inline libcxx/utils/data/ignore_format.txt:884]: Can you run `clang-format` on the new test you added instead? We would like to avoid adding new entries to the `ignore_format.txt` file -- we're trying to get rid of it.

File: libcxx/include/__random/uniform_real_distribution.h
Comment (ldionne) [inline libcxx/include/__random/uniform_real_distribution.h:31]: We have a utility called `__libcpp_random_is_valid_urng` in [libcxx/include/__random/is_valid.h](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__random/is_valid.h#L57-L61) -- should we instead create something similar to it and handle http://eel.is/c++draft/rand.req.genl more generally?

It seems like the restriction for `is_floating_point` is shared by many distributions. From looking at the synopsis I can find:

```
template<class RealType = double>
class uniform_real_distribution;

template<class RealType = double>
class exponential_distribution;

template<class RealType = double>
class gamma_distribution;

template<class RealType = double>
class weibull_distribution;

template<class RealType = double>
class extreme_value_distribution;

template<class RealType = double>
class normal_distribution;

template<class RealType = double>
class lognormal_distribution;

template<class RealType = double>
class chi_squared_distribution;

template<class RealType = double>
class cauchy_distribution;

template<class RealType = double>
class fisher_f_distribution;

template<class RealType = double>
class student_t_distribution;

template<class RealType = double>
class piecewise_constant_distribution;

template<class RealType = double>
class piecewise_linear_distribution;
```

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/__random/uniform_real_distribution.h
Comment (NhatNguyen1810) [reply libcxx/include/__random/uniform_real_distribution.h:31]: @ldionne Ide
are you suggesting to replace `is_floating_point` by `__libcpp_random_is_valid_urng` in static assert?

**Comment by NhatNguyen1810** [comment]
@ldionne I
looks like there is an issue with the CI build, can you take a look? the command I ran was `clang-format -style=llvm [file]`. But before that, when you triggered the last build, I also got an error around the ignore-format.txt file, the error msg was confusing too. 

Whats your suggestion for a fix to this?

**Comment by AdvenamTacet** [comment]
@NhatNguyen1810 you can see error report by clicking details next to the buildkite and then on the job which failed. It looks like your code works with GCC only. You can find details here: https://buildkite.com/llvm-project/libcxx-ci/builds/31338#018b8dda-3da3-4049-af36-d3bbe1bf054c

`FAIL: llvm-libc++-shared.cfg.in :: libcxx/random/random.uniform.real/traits_mismatch.verify.cpp (1664 of 9610)` - that's the test which failed.

```bash
# .---command stderr------------
--
  | # \| error: 'expected-error' diagnostics seen but not expected:
  | # \|   File /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-6be6c9a4344c-1/llvm-project/libcxx-ci/build/generic-cxx26/include/c++/v1/__random/uniform_real_distribution.h Line 31: static assertion failed due to requirement 'is_floating_point<int>::value': result_type must be floating type
  | # \|   File /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-6be6c9a4344c-1/llvm-project/libcxx-ci/libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp Line 20: cannot find end ('}}') of expected string
  | # \| 2 errors generated.
  | # `-----------------------------
```

**Comment by NhatNguyen1810** [comment]
> @NhatNguyen1810 you can see error report by clicking details next to the buildkite and then on the job which failed. It looks like your code works with GCC only. You can find details here: https://buildkite.com/llvm-project/libcxx-ci/builds/31338#018b8dda-3da3-4049-af36-d3bbe1bf054c
> 
> `FAIL: llvm-libc++-shared.cfg.in :: libcxx/random/random.uniform.real/traits_mismatch.verify.cpp (1664 of 9610)` - that's the test which failed.
> 
> ```shell
> # .---command stderr------------
> --
>   | # \| error: 'expected-error' diagnostics seen but not expected:
>   | # \|   File /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-6be6c9a4344c-1/llvm-project/libcxx-ci/build/generic-cxx26/include/c++/v1/__random/uniform_real_distribution.h Line 31: static assertion failed due to requirement 'is_floating_point<int>::value': result_type must be floating type
>   | # \|   File /home/libcxx-builder/.buildkite-agent/builds/google-libcxx-builder-6be6c9a4344c-1/llvm-project/libcxx-ci/libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp Line 20: cannot find end ('}}') of expected string
>   | # \| 2 errors generated.
>   | # `-----------------------------
> ```

Hi, thanks for answering me! Yeah I saw that error message, but I guess its an error because I have multiple lines for the error message? 

To give you more contex, originally I ran `git-clang-format`, then the CI passed the buildkite, however it failed the last step with some info about the `ignore-format.txt` file. I added the test file to the `ignore-format.txt` file and still got that error. 

@ldionne Ide
suggested me to run `clang-format` on the test file. I ran `clang-format -style=llvm` on the test file, then it became like above and the CI build breaks. Now I am not sure how to move forward. Any ideas? Thanks!!

**Comment by changkhothuychung** [comment]
@philnik777 phil
can you take a look when you have time? thanks.

File: libcxx/include/__random/uniform_real_distribution.h
Comment (ldionne) [reply libcxx/include/__random/uniform_real_distribution.h:31]: Add the following to `libcxx/include/__random/is_valid.h`:

```
// [rand.req.genl]/1.4:
// The effect of instantiating a template that has a template type parameter
// named RealType is undefined unless the corresponding template argument is
// cv-unqualified and is one of float, double, or long double.

template<class> struct __libcpp_random_is_valid_realtype : false_type {};
template<> struct __libcpp_random_is_valid_realtype<float> : true_type {};
template<> struct __libcpp_random_is_valid_realtype<double> : true_type {};
template<> struct __libcpp_random_is_valid_realtype<long double> : true_type {};
```

Then, in the distributions mentioned above, add the following `static_assert`:

```
static_assert(__libcpp_random_is_valid_realtype<_RealType>::value, "RealType must be a supported floating-point type");
```

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp
Comment (mordante) [inline libcxx/test/libcxx/random/random.uniform.real/traits_mismatch.verify.cpp:20]:
```suggestion
// expected-error@*:* {{static assertion failed due to requirement 'is_floating_point<int>::value': result_type must be floating type}}
struct test_random : public std::uniform_real_distribution<int> {};
```
This is the typical way we write these validation types. Maybe add a few more invalid types.

**Comment by changkhothuychung** [comment]
@mordante mordanta
there are timed out and failed tests that are not related to my PR at all. Can you take a look and let me know how to fix this? Thanks.

**Comment by AdvenamTacet** [comment]
It's a known problem at the moment. There is no need to take any action at the moment.

**Review by ldionne — APPROVED** [review]
LGTM with latest changes. I will merge once CI is done.

**Comment by changkhothuychung** [comment]
@ldionne ld
there was an error with code formatter, I made a push to fix it. Does everything look good to be merged now? Can you help me merge? Thanks.

**Review by EricWF — CHANGES_REQUESTED** [review]
Please give me 24 hours to test this change at Google to see what the impact is.

**Comment by ldionne** [comment]
@changkhothuychung You have a setting on your Github account to "hide your email". Please disable it, otherwise upon committing this the email address will be `35616780+changkhothuychung@users.noreply.github.com`. We're trying to prevent that from happening for commit attribution purposes.

**Comment by changkhothuychung** [comment]
> @changkhothuychung You have a setting on your Github account to "hide your email". Please disable it, otherwise upon committing this the email address will be `35616780+changkhothuychung@users.noreply.github.com`. We're trying to prevent that from happening for commit attribution purposes.

Do you see my email now? it should show up as nhat7203@gmail.com, thanks!

**Comment by ldionne** [comment]
Yes I can see it now, thanks.

@EricWF Do you have any results internally? We've been sitting on this for a while and I'd like to merge it by EOD if possible, since it looks otherwise good to go.

**Comment by EricWF** [comment]
@ldionne I have to run the tests off hours, they're set up to run tonight.

**Comment by ldionne** [comment]
Sounds good, I'll check again tomorrow morning or mid-day. I'm eager to close this PR :)

**Comment by ldionne** [comment]
> @ldionne For one reason or another, the test runs got aborted last night. I'm rerunning them again tonight.

Do you have any news about this?

**Comment by EricWF** [comment]
Yes, there are a few users in Google3. I would like to clean them up before this lands, but that will take until Wednesday. None of them seem show-stopping though. All of the usages are weird, but otherwise fine, code.

If you're so inclined,  I would love the extra time. Otherwise, this LGTM.

**Review by EricWF — APPROVED** [review]
I would ask kindly that you wait a day or two, but that's only a request.

LGTM.

**Comment by ldionne** [comment]
Cool, thanks a bunch for the information. I'll land this on Thursday. This is not an emergency at all but it would be good to get it done. Hopefully that should give you enough time.

**Pull request merged**
